.github/workflows/CI.yaml

```diff
@@ -1,38 +1,174 @@
-name: Kernel CI
- 
+name: CI
+
 on:
   push:
     branches: ["main"]
   pull_request:
-    types: [opened, synchronize]
- 
+    types: [opened, reopened, synchronize]
+
 jobs:
+  setup:
+    name: Install dependencies
+    timeout-minutes: 15
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v4
+
+      - uses: pnpm/action-setup@v4
+
+      - name: Setup Node.js environment
+        uses: actions/setup-node@v4
+        with:
+          node-version: 22
+          cache: "pnpm"
+
+      - name: Install dependencies
+        run: pnpm install
+
   build:
-    name: Build and Test
+    name: Build
     timeout-minutes: 15
     runs-on: ubuntu-latest
- 
+    needs: [setup]
     steps:
       - name: Check out code
         uses: actions/checkout@v4
+
+      - uses: pnpm/action-setup@v4
+
+      - name: Cache turbo build setup
+        uses: actions/cache@v4
         with:
-          fetch-depth: 2
- 
-      - uses: pnpm/action-setup@v3
+          path: .turbo
+          key: ${{ runner.os }}-turbo-build-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-turbo-build-
+
+      - name: Setup Node.js environment
+        uses: actions/setup-node@v4
+        with:
+          node-version: 22
+          cache: "pnpm"
+
+      - name: Install dependencies
+        run: pnpm install
+
+      - name: Generate prisma client
+        working-directory: apps/backend
+        run: pnpm run prisma:generate
+
+      - run: pnpm build
+
+  lint:
+    name: Lint
+    timeout-minutes: 15
+    runs-on: ubuntu-latest
+    needs: [setup]
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v4
+
+      - uses: pnpm/action-setup@v4
+
+      - name: Cache turbo build setup
+        uses: actions/cache@v4
+        with:
+          path: .turbo
+          key: ${{ runner.os }}-turbo-lint-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-turbo-lint-
+
+      - name: Setup Node.js environment
+        uses: actions/setup-node@v4
+        with:
+          node-version: 22
+          cache: "pnpm"
+
+      - name: Install dependencies
+        run: pnpm install
+
+      - run: pnpm lint --parallel --continue
+        id: run_lint
+
+  format:
+    name: Check format
+    timeout-minutes: 15
+    runs-on: ubuntu-latest
+    needs: [setup]
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v4
+
+      - uses: pnpm/action-setup@v4
+
+      - name: Setup Node.js environment
+        uses: actions/setup-node@v4
+        with:
+          node-version: 22
+          cache: "pnpm"
+
+      - name: Install dependencies
+        run: pnpm install
+
+      - run: pnpm format:check
+
+  test:
+    name: Test kernel
```

Comment:
To save resources, Test should be skipped unless there is a change in the `packeges/kernel` or `pnpm-lock.yaml` and some config files. https://github.com/dorny/paths-filter would be useful.

Reply:
done :+1: